src/hotspot/share/opto/loopnode.cpp

```diff
@@ -3675,9 +3679,16 @@ void PhaseIdealLoop::replace_parallel_iv(IdealLoopTree *loop) {
         incr2->in(1)->uncast() != phi2 ||
         incr2 == incr ||
         incr2->Opcode() != Op_AddI ||
-        !incr2->in(2)->is_Con())
+        !incr2->in(2)->is_Con()) {
       continue;
+    }
 
+    if (incr2->in(1)->is_ConstraintCast() && !incr2->in(1)->in(0)->is_RangeCheck()) {
+      // The form of Phi->CastXX->AddX appears when using Preconditions.checkIndex, and it would
+      // be recognized as additional IV when 1) Phi != phi2, 2) CastXX is controlled by RangeCheck
+      // (to reflect changes in LibraryCallKit::inline_preconditions_checkIndex)
```

Comment:
Thank you for updating comment. But it still do not say which issue you are trying to avoid. Yes, this pattern can be generated by inline_preconditions_checkIndex.
Currently your check skip search for parallel IV if CastII is NOT controlled by RangeCheck!
But the comment in the BODY of check talks about using parallel IV.
It is contradicting.
Aslo `1) Phi != phi2` says nothing to me - your check does not have such condition.

Reply:
Why not just say that we should skip `AddI->CastII->Phi` case if CastII is not controlled by local RangeCheck.

Reply:
> Why not just say that we should skip `AddI->CastII->Phi` case if CastII is not controlled by local RangeCheck.

Yes, this is exactly what I want to say. Maybe I should update comments.